src/main/java/io/confluent/connect/jdbc/sink/JdbcSinkTask.java

```diff
@@ -88,28 +103,48 @@ public void put(Collection<SinkRecord> records) {
       }
       SQLException sqlAllMessagesException = new SQLException(sqleAllMessages);
       sqlAllMessagesException.setNextException(sqle);
-      if (remainingRetries == 0) {
-        log.error(
-            "Failing task after exhausting retries; "
-              + "encountered {} exceptions on last write attempt. "
-              + "For complete details on each exception, please enable DEBUG logging.",
-            totalExceptions);
-        int exceptionCount = 1;
-        for (Throwable e : sqle) {
-          log.debug("Exception {}:", exceptionCount++, e);
-        }
-        throw new ConnectException(sqlAllMessagesException);
-      } else {
+      if (remainingRetries > 0) {
         writer.closeQuietly();
         initWriter();
         remainingRetries--;
         context.timeout(config.retryBackoffMs);
         throw new RetriableException(sqlAllMessagesException);
+      } else {
+        if (reporter != null) {
+          unrollAndRetry(records);
+        } else {
+          log.error(
+              "Failing task after exhausting retries; "
+                  + "encountered {} exceptions on last write attempt. "
+                  + "For complete details on each exception, please enable DEBUG logging.",
+              totalExceptions);
+          int exceptionCount = 1;
+          for (Throwable e : sqle) {
+            log.debug("Exception {}:", exceptionCount++, e);
+          }
+          throw new ConnectException(sqlAllMessagesException);
+        }
       }
     }
     remainingRetries = config.maxRetries;
   }
 
+  private void unrollAndRetry(Collection<SinkRecord> records) {
```

Comment:
Likewise when this method is called, connection has failed transaction and all subsequent record would end up in dlq.

Reply:
I've added `closeQuietly()` calls and an integration test to test this case. Let me know what you think.

Reply:
@aakashnshah Thank you for adding closeQuitely(). I think when there is SQL exception in first batch here https://github.com/confluentinc/kafka-connect-jdbc/pull/999/files/e2f6a063b46ce417c0727fb29af41ceab7081710#diff-70ff6a1afa83a978851964510faa22330153591783ddf054620daac2b9862fdeR85 connection will be in bad state. As there is no closeQuietly before calling write statement here https://github.com/confluentinc/kafka-connect-jdbc/pull/999/files/e2f6a063b46ce417c0727fb29af41ceab7081710#diff-70ff6a1afa83a978851964510faa22330153591783ddf054620daac2b9862fdeR136 first record of the batch will always end up in dlq independent of wether it is good or bad record.

Reply:
@aakashnshah I think this still needs to be addressed? FWICT @prakashrajpoudel is correct as there is no call to `JdbcWriter::closeQuietly` before the first call to `JdbcWriter::write` in `JdbcSinkTask::unrollAndRetry`. We might consider adding an initial call to `JdbcWriter::closeQuietly` to the beginning of `JdbcSinkTask::unrollAndRetry`.

Reply:
@C0urante @prakashrajpoudel That is a very good point. I've added the `JdbcWriter::closeQuietly` call to the first line in `unrollAndRetry`, so that any time it is called, we first call this method to ensure the record doesn't fail from the failed transaction. LMK what you think.

Reply:
Looks good to me 👍 

Could we add a comment above that line explaining why we preemptively close the writer? I'd definitely be scratching my head over that if I saw it without this context.

Reply:
Looks good to me 👍 

 It seems like when this `closeQuietly` is called, it eventually not only close connection  but cached connection provider. Is it different than connection pool ? 🤷‍♂️  I would really appreciate if someone can explain me what this is
https://github.com/confluentinc/kafka-connect-jdbc/blob/master/src/main/java/io/confluent/connect/jdbc/util/CachedConnectionProvider.java#L114